File: tt-api/src/Domain/Entities/Decant.cs
using Domain.ValueObjects;

namespace Domain.Entities;

public class Decant
{
    public string JobNo { get; set; } = null!;
    public string CustomerCode { get; set; } = null!;
    public string CustomerName { get; set; } = null!;
    public string WhsCode { get; set; } = null!;
    public string? ReferenceNo { get; set; }
    public string? Remark { get; set; }
    public DecantStatus Status { get; set; } = null!;
    public string CreatedBy { get; set; } = null!;
    public DateTime CreatedDate { get; set; }
    public string? CompletedBy { get; set; }
    public DateTime? CompletedDate { get; set; }
    public string? CancelledBy { get; set; }
    public DateTime? CancelledDate { get; set; }
    // TODO is it a concious choice of creating an aggregate? Perhaps it could be differentiate from other entities?
    public ICollection<DecantItem> Items { get; set; } = new List<DecantItem>();

    public bool CanComplete => CanEdit;
    public bool CanEdit
        => !Status.Equals(DecantStatus.Cancelled) && !Status.Equals(DecantStatus.Completed);

    public void Complete(string completedBy, DateTime completedDate)
    {
        CompletedBy = completedBy;
        CompletedDate = completedDate;
        Status = DecantStatus.Completed;
    }

    public void Cancel(string cancelledBy, DateTime cancelledDate)
    {
        CancelledBy = cancelledBy;
        CancelledDate = cancelledDate;
        Status = DecantStatus.Cancelled;
    }

    public DecantItem AddDecantItem(Pallet pallet, ICollection<int> newQuantities)
    {
        if (Items.Where(i => i.SourcePalletId == pallet.Id).FirstOrDefault() != null)
            throw new InvalidOperationException();

        var newItem = new DecantItem
        {
            SourcePalletId = pallet.Id,
            SourceQty = pallet.Qty
        };
        int seqenceNo = 0;
        foreach (var qty in newQuantities)
        {
            var decantItemPallet = new DecantItemPallet
            {
                Qty = qty,
                SequenceNo = ++seqenceNo,
            };
            decantItemPallet.CopyDataFromPallet(pallet);
            newItem.NewPallets.Add(decantItemPallet);
        }
        Items.Add(newItem);
        return newItem;
    }
}

File: ttlogix-web/App/ClientApp/src/store/models/decant.ts
export class Decant {
    jobNo: string;
    customerCode: string;
    customerName: string;
    whsCode: string;
    referenceNo: string;
    remark: string;
    status: number;
    createdBy: string;
    createdDate: string;
    completedBy: string;
    completedDate: string;
    cancelledBy: string;
    cancelledDate: string;
    items: Array<DecantPallet>;
    canComplete: boolean;
    canEdit: boolean;

    // Local properties
    _detailsList: Array<DecantDetailsList>;
    get detailsList() {
        if(!this._detailsList) {
            this._detailsList = [];
            if(this.items && this.items.length > 0){
                this.items.forEach((pallet) => { 
                    if(pallet.newPallets && pallet.newPallets.length > 0){
                        pallet.newPallets.forEach((newPallet) => { this._detailsList.push(new DecantDetailsList(newPallet, pallet.sourcePalletId, pallet.sourceQty)) });
                    }
                }
            )}
        }
        return this._detailsList;
    }
    set detailsList(list:Array<DecantDetailsList>){
        this._detailsList = list;
    }
}

export class DecantPallet {
    sourcePalletId: string;
    sourceQty: number;
    newPallets: Array<DecantNewPallet>;
}

export class DecantNewPallet {
    palletId: string;
    qty: number;
    sequenceNo: number;
    productCode: string;
    supplierId: string;
    length: number;
    width: number;
    height: number;
    netWeight: number;
    grossWeight: number;
}

export class DecantDetailsList {
    constructor(newPallet:DecantNewPallet, origPID: string, origQty: number){
        this.originalPID = origPID;
        this.productCode = newPallet.productCode;
        this.supplierID = newPallet.supplierId;
        this.originalQty = origQty;
        this.sequenceNo = newPallet.sequenceNo;
        this.pid = newPallet.palletId;
        this.newQty = newPallet.qty;
    }

    originalPID: string;
    productCode: string;
    supplierID: string;
    originalQty: number;
    sequenceNo: number;
    pid: string;
    newQty: number;
}